examples/fluids/qfunctions/eulervortex.h

```diff
@@ -386,14 +547,92 @@ CEED_QFUNCTION(IFunction_Euler)(void *ctx, CeedInt Q,
     // ---- rho (u x u) + P I3
     for (int j=0; j<3; j++)
       for (int k=0; k<3; k++)
-        dv[k][j+1][i]  -= wdetJ*((rho*u[j]*u[0] + (j==0?P:0))*dXdx[k][0] +
-                                 (rho*u[j]*u[1] + (j==1?P:0))*dXdx[k][1] +
-                                 (rho*u[j]*u[2] + (j==2?P:0))*dXdx[k][2]);
+        dv[k][j+1][i]  -= wdetJ*((rho*u[j]*u[0] + (j==0?P:0.))*dXdx[k][0] +
+                                 (rho*u[j]*u[1] + (j==1?P:0.))*dXdx[k][1] +
+                                 (rho*u[j]*u[2] + (j==2?P:0.))*dXdx[k][2]);
     // -- Total Energy Density
     // ---- (E + P) u
     for (int j=0; j<3; j++)
       dv[j][4][i]  -= wdetJ * (E + P) * (u[0]*dXdx[j][0] + u[1]*dXdx[j][1] +
                                          u[2]*dXdx[j][2]);
+
+    // -- Stabilization terms
+    // ---- jacob_F_conv[3][5][5] = dF(convective)/dq at each direction
+    CeedScalar jacob_F_conv[3][5][5] = {{{0.}}};
+    computeFluxJacobian_Euler(jacob_F_conv, rho, u, E, gamma);
+
+    // ---- Transpose of the Jacobian
+    CeedScalar jacob_F_conv_T[3][5][5];
+    for (int j=0; j<3; j++)
+      for (int k=0; k<5; k++)
+        for (int l=0; l<5; l++)
+          jacob_F_conv_T[j][k][l] = jacob_F_conv[j][l][k];
+
+    // ---- dqdx collects drhodx, dUdx and dEdx in one vector
+    CeedScalar dqdx[5][3];
+    for (int j=0; j<3; j++) {
+      dqdx[0][j] = drhodx[j];
+      dqdx[4][j] = dEdx[j];
+      for (int k=0; k<3; k++)
+        dqdx[k+1][j] = dUdx[k][j];
+    }
+
+    // ---- strong_conv = dF/dq * dq/dx    (Strong convection)
+    CeedScalar strong_conv[5] = {0.};
+    for (int j=0; j<3; j++)
+      for (int k=0; k<5; k++)
+        for (int l=0; l<5; l++)
+          strong_conv[k] += jacob_F_conv[j][k][l] * dqdx[l][j];
+
+    // ---- Strong residual
+    CeedScalar strong_res[5];
+    for (int j=0; j<5; j++)
+      strong_res[j] = q_dot[j][i] + strong_conv[j];
+
+    // ---- Tau elements
+    CeedScalar uX[3];
+    for (int j=0; j<3; j++)
+      uX[j] = dXdx[j][0]*u[0] + dXdx[j][1]*u[1] + dXdx[j][2]*u[2];
+    const CeedScalar uiujgij = uX[0]*uX[0] + uX[1]*uX[1] + uX[2]*uX[2];
+    const CeedScalar Cc      = 1.;
+    const CeedScalar Ce      = 1.;
+    const CeedScalar f1      = rho * sqrt(uiujgij);
+    const CeedScalar TauC    = (Cc * f1) /
+                               (8. * (dXdxdXdxT[0][0] + dXdxdXdxT[1][1] + dXdxdXdxT[2][2]));
+    const CeedScalar TauM    = 1. / (f1>1. ? f1 : 1.);
+    const CeedScalar TauE    = TauM / (Ce * cv);
+    const CeedScalar Tau[5]  = {TauC, TauM, TauM, TauM, TauE};
```

Comment:
I want to think of all the tau terms as having units of time, but these apparently don't. I recall discussing this formulation (or something quite like it), but I think we didn't write down how we got these expressions.

Reply:
I am writing this for posterity. Yes, those terms were taken from Whiting et al 2003 which are for primitive variables. Thanks for the fix!